src/main/java/io/cryostat/recordings/RecordingTargetHelper.java

```diff
@@ -365,6 +438,42 @@ public IRecordingDescriptor startRecording(
 
         return builder.build();
     }
+
+    private void scheduleRecordingStopNotification(
+            String recordingName, long delay, ConnectionDescriptor connectionDescriptor) {
+        ScheduledFuture<Optional<IRecordingDescriptor>> scheduledFuture =
+                this.scheduler.schedule(
+                        () -> {
+                            return targetConnectionManager.executeConnectedTask(
+                                    connectionDescriptor,
+                                    connection -> {
+                                        Optional<IRecordingDescriptor> desc =
+                                                getDescriptorByName(connection, recordingName);
+
+                                        long recordingStopped =
+                                                desc.stream()
+                                                        .map(IRecordingDescriptor::getState)
+                                                        .filter(
+                                                                s ->
+                                                                        s.equals(
+                                                                                RecordingState
+                                                                                        .STOPPED))
+                                                        .count();
+                                        if (recordingStopped > 0) {
+                                            this.notifyRecordingStopped(
+                                                    recordingName,
+                                                    connectionDescriptor.getTargetId());
+                                        }
+
+                                        return desc;
+                                    });
+                        },
+                        delay + TIMESTAMP_DRIFT_SAFEGUARD,
+                        TimeUnit.MILLISECONDS);
+
+        scheduledStopNotifications.put(recordingName, scheduledFuture);
```

Comment:
I think the notifications need to be mapped not only by the recording name but also the name of the associated target, since names are only unique within a target. So either this should be some computed key that can be unambiguously taken back apart into a targetId part and a recordingName part, or it should be a `Map<TargetId<Map<RecordingName, ScheduledFuture>>>`

Reply:
Are TargetIds unique?

Reply:
Yes - if two targetIds are the same then they are the same JMX service URL, or other future identifier/URL for a target.

It is possible that one target can be referred to by two or more different targetIds, however. For example, when you use our standard development setup with JDP, Cryostat can discover itself there with the JMX service URL `cryostat:9091`. You can then add a custom target for `localhost:0`, which tells the JVM to connect to itself locally. There are various other ways such a scenario can arise, like having multiple DNS records for different hostnames pointing to the same IP where a JVM is listening.

I think in those cases, at least for now, that'll be a bit of a caveat emptor. To properly fix that we would need to actually perform DNS queries whenever a targetId is discovered or added and ensure that it's really referring to a distinct JVM from any others we already know about, and in the future we might also need to know how to map JMX service URLs to Jolokia URLs etc.

I think this would only ever present a "problem" if a client started a recording `r1` using `targetId1` that refers to `jvmA`, and then deleted `r1` using `targetId2` that also refers to `jvmA`, or other strange scenarios like that. In this case, we would end up deleting the recording but not the scheduled notification, so we would waste a bit of time/energy in the future connecting back to `targetId1` and checking if `r1` is still there, even though we "should" know that we've already deleted it. The notification still wouldn't be emitted in the end, so I don't think it's really a real problem.

Reply:
Sounds good, thanks